src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,156 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> tail;
+    private Node<T> head;
+    private int size;
+
+    private static class Node<T> {
+        private T element;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T element, Node<T> next) {
+            this.element = element;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            head = new Node<>(null, value, null);
+            tail = head;
+        } else {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, head);
+            head = newNode;
+            head.next.prev = newNode;
+        } else {
+            Node<T> nodeOfIndex = getNodeOfIndex(index);
+            Node<T> newNode = new Node<>(nodeOfIndex.prev, value, nodeOfIndex);
+            nodeOfIndex.prev.next = newNode;
+            nodeOfIndex.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeOfIndex(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (checkIndex(index)) {
+            T oldValue = getNodeOfIndex(index).element;
+            getNodeOfIndex(index).element = value;
+            return oldValue;
+        }
+        throw new IndexOutOfBoundsException();
```

Comment:
nw when you're throwing exception inside the check index method we don't have to throw it here and we don't need this if block. this will work because if index is invalid the exception will be thrown and the method will be stopped

```suggestion
        checkIndex(index);
        T oldValue = getNodeOfIndex(index).element;
        getNodeOfIndex(index).element = value;
        return oldValue;
```